base/pkg/entry.jl

```diff
@@ -352,19 +359,27 @@ function update(branch::AbstractString)
         end
     end
     instd = Read.installed(avail)
-    free = Read.free(instd)
+    free  = Read.free(instd)
     for (pkg,ver) in free
-        Cache.prefetch(pkg, Read.url(pkg), [a.sha1 for (v,a)=avail[pkg]])
+        Cache.prefetch(pkg, Read.url(pkg), [a.sha1 for (v,a) in avail[pkg]])
     end
     fixed = Read.fixed(avail,instd)
     for (pkg,ver) in fixed
         ispath(pkg,".git") || continue
         with(GitRepo, pkg) do repo
-            if LibGit2.isattached(repo) && !LibGit2.isdirty(repo)
-                info("Updating $pkg...")
-                @recover begin
-                    LibGit2.fetch(repo)
-                    LibGit2.merge!(repo, fastforward=true)
+            if LibGit2.isattached(repo)
+                if LibGit2.isdirty(repo)
+                    warn("Package $pkg: skipping update (dirty)...")
+                else
+                    prev_sha = string(LibGit2.head_oid(repo))
+                    @recover begin
+                        LibGit2.fetch(repo)
+                        LibGit2.merge!(repo, fastforward=true)
+                    end
+                    post_sha = string(LibGit2.head_oid(repo))
+                    branch = LibGit2.branch(repo)
+                    info("Updating $pkg $branch...",
+                          prev_sha != post_sha ? " $(prev_sha[1:8]) → $(post_sha[1:8])" : "")
```

Comment:
This might be a little confusing, since any errors that get printed the `@recover` block above will occur *before* the updating message.